ext/standard/string.c

```diff
@@ -2751,56 +2753,152 @@ PHP_FUNCTION(quotemeta)
 }
 /* }}} */
 
-/* {{{ proto int ord(string character)
-   Returns ASCII value of character
+/* {{{ proto int ord(string character[, string encoding])
+   Returns code point of character
    Warning: This function is special-cased by zend_compile.c and so is bypassed for constant string argument */
 PHP_FUNCTION(ord)
 {
-	char   *str;
-	size_t str_len;
+  char   *str;
+  size_t str_len;
+  char   *enc = NULL;
+  size_t enc_len;
+  enum entity_charset charset;
+  long cp = 0;
+  size_t pos = 0;
+  int status;
 
 #ifndef FAST_ZPP
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &str, &str_len) == FAILURE) {
-		return;
-	}
+  if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|s", &str, &str_len, &enc, &enc_len) == FAILURE) {
+    return;
+  }
 #else
-	ZEND_PARSE_PARAMETERS_START(1, 1)
-		Z_PARAM_STRING(str, str_len)
-	ZEND_PARSE_PARAMETERS_END();
+  ZEND_PARSE_PARAMETERS_START(1, 2)
+    Z_PARAM_STRING(str, str_len)
+    Z_PARAM_OPTIONAL
+    Z_PARAM_STRING(enc, enc_len)
+  ZEND_PARSE_PARAMETERS_END();
 #endif
 
-	RETURN_LONG((unsigned char) str[0]);
+  charset = enc == NULL ? cs_8859_1 : determine_charset(enc);
+
+  cp = get_next_char(
+    charset, (const unsigned char*) str, str_len, &pos, &status
+  );
+
+  if (status == FAILURE) {
+    cp = charset == cs_utf_8 ? 0xfffd : 0x3f;
+  }
+
+  RETURN_LONG(cp);
 }
 /* }}} */
 
-/* {{{ proto string chr(int ascii)
-   Converts ASCII code to a character
+/* {{{ proto string chr(int codepoint[, string encoding])
+   Converts codepoint to a character
    Warning: This function is special-cased by zend_compile.c and so is bypassed for constant integer argument */
 PHP_FUNCTION(chr)
 {
-	zend_long c;
+  zend_long c;
+  zend_string* enc = NULL;
+	enum entity_charset charset;
+	zend_string* buf;
+  unsigned int cp;
+	size_t pos = 0;
+	int status;
+
+  if (1 > ZEND_NUM_ARGS()) {
+    WRONG_PARAM_COUNT;
```

Comment:
Although unrelated to this patch, I wonder why we use ZEND_PARSE_PARAMS_QUIET for chr(). I don't remember any legacy reasons to why this might be it